design/proposal-controller-code-generation.md

```diff
@@ -0,0 +1,129 @@
+
+### Merge Patching / Injection / Overlays
+The generation of CRD types and reconciliation methods are challenging when the cloud provider interactions don’t map 1:1 with the CRD types. Instead of the kubebuilder paradigm of generating source code files that users modify, it would be interesting to explore injection points where some of the components outlined in “Output Targets” could be replaced or augmented, possibly in a middleware model. The code generation process should be modular and pluggable, allowing for customizations to the rendered code. Knowing exactly what form these overlays will take is going to require first prototyping the MRS->Go tools, but here are a few thoughts on what this could look like.
+
+It’s possible that we’ll wind up with injection points in the Frontend and Backend of the generation pipeline; manipulating CRD structure in the Frontend, and customizing code generation in the Backend. For instance in the Frontend, the json-schema object structure can be directly manipulated as in-memory go structs or json. Driving code generation by manipulating the MRS structure could be less fragile than directly working with the code generation routines in the backend. For a problem like customizing Reference resolution, we probably also need backend hooks where user-supplied code can be directly invoked.
+
+Alternatively, the generated code could follow a plugin registration pattern (keyed to a particular provider+resource+component) where developers can override/augment/append generated code without touching the generic code generation functions. Given that triggering code generation could have unintended side effects when provider metadata descriptions are revised, this could give developers more control to modify controllers without worrying about broader versioning concerns.
+
+### Unstructured data
+Provider types can have cross-type relationships, for instance a subnetwork in GCP belongs to a VPC. The reference in the APIs can be represented as a simple string, with no “strongly typed” relationship in the generated code. As we assess Provider API discovery documents, we should observe whether these relationships are represented in a structured way that could drive code generation to build [Reference Resolvers]. It’s possible that the resolution logic is complex enough to move into Overlays, and all we need to solve at this stage is making sure we preserve metadata from the provider in the MRS.
+
+### Deprecate Kubebuilder
+Most of the interesting code generation we get from Kubebuilder actually seems to come from [controller-tools]. Kubebuilder is more geared towards generating boilerplate, which will be superseded by the boilerplate generated by this project. We will likely want to remove `kubebuilder` from our workflow and directly work with [controller-tools]/[controller-runtime], building from the annotated CRD structure instead of special comments in go code.
```

Comment:
Do we currently use `kubebuilder` anywhere? I think it has mostly already been removed (except `crossplane-cli`)

Reply:
I think it lives on as a documented practice in the services developer guide, but isn't involved in any of our code generation pipelines in practice.